### ISSUE TITLE
Unexpected behaviour of `hipMemcpy3D`

### ISSUE THREAD

**Issue by luraess** [ROCm-Developer-Tools/HIP#3289]
I am trying to implement `hipMemcpy3D` function in a finite-difference stencil halo exchange code. I am encountering an issue in defining `srcPos`. From the docs and by analogy with CUDA, if srcPos[1] == 0, then srcPos[2] and srcPos[3] also equal to 0, otherwise one need to always pass srcPos[1] = sizeof(DAT). Applying this logic works but results in a shift by 1 between the copied and expected results.

Running the MWE hereafter:
```cpp
#define __HIP_PLATFORM_AMD__
#include <hip/hip_runtime.h>

#include <stdio.h>
#include <stdlib.h>

#define DAT double

#define hip_check(ex)                                                          \
  do {                                                                         \
    hipError_t status = ex;                                                    \
    if (status != hipSuccess) {                                                \
      fprintf(stderr, "Error: %s at %s:%d\n", #ex, __FILE__, __LINE__);        \
      const char *msg = hipGetErrorString(status);                             \
      fprintf(stderr, "%s\n", msg);                                            \
    }                                                                          \
  } while (0)

int main() {
  int nx = 7, ny = 5;
  int sl = 2;
  DAT *P = nullptr, *P_h, *buf = nullptr, *buf_d = nullptr;

  buf = (DAT *)malloc(nx * sizeof(DAT));
  memset(buf, 0, nx * sizeof(DAT));
  hip_check(hipHostRegister(buf, nx * sizeof(DAT), hipHostRegisterMapped));
  hip_check(hipHostGetDevicePointer((void **)(&buf_d), buf, 0));

  P_h = (DAT *)malloc(nx * ny * sizeof(DAT));
  for (int iy = 0, idx = 0; iy < ny; ++iy) {
    for (int ix = 0; ix < nx; ++ix, ++idx) {
      P_h[idx] = iy * 10 + ix;
    }
  }

  hip_check(hipMalloc(&P, nx * ny * sizeof(DAT)));
  hip_check(hipMemcpy(P, P_h, nx * ny * sizeof(DAT), hipMemcpyHostToDevice));

  hipMemcpy3DParms params = {
      .srcPos = make_hipPos(sizeof(DAT), sl, 0),
      .srcPtr = make_hipPitchedPtr(P, nx * sizeof(DAT), nx, ny),
      .dstPos = make_hipPos(0, 0, 0),
      .dstPtr = make_hipPitchedPtr(buf_d, nx * sizeof(DAT), nx, 1),
      .extent = make_hipExtent(nx * sizeof(DAT), 1, 1),
      .kind = hipMemcpyDeviceToHost,
  };

  hip_check(hipMemcpy3D(&params));

  printf("Array P_h: [ ");
  for (int ix = 0; ix < nx; ++ix) {
    printf("%lf ", P_h[ix + sl*nx]);
  }
  printf("]\n");

  printf("Array buf: [ ");
  for (int ix = 0; ix < nx; ++ix) {
    printf("%lf ", buf[ix]);
  }
  printf("]\n");

  return 0;
}
```
returns following results for `srcPos = make_hipPos(sizeof(DAT), sl, 0)` where `sl = 2`:
```
lurass@nid005075:~/scratch/test> ./a.out
Array P_h: [ 20.000000 21.000000 22.000000 23.000000 24.000000 25.000000 26.000000 ]
Array buf: [ 21.000000 22.000000 23.000000 24.000000 25.000000 26.000000 30.000000 ]
```

Defining `srcPos = make_hipPos(0, sl, 0)` would produce the expected results (here for `sl = 2`):
```
lurass@nid005075:~/scratch/test> ./a.out
Array P_h: [ 20.000000 21.000000 22.000000 23.000000 24.000000 25.000000 26.000000 ]
Array buf: [ 20.000000 21.000000 22.000000 23.000000 24.000000 25.000000 26.000000 ]
```
but **fail** for `sl = 1`:
```
lurass@nid005075:~/scratch/test> ./a.out
invalid argument
Array P_h: [ 10.000000 11.000000 12.000000 13.000000 14.000000 15.000000 16.000000 ]
Array buf: [ 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000 0.000000 ]
```

Am I doing something wring or is there indeed a bug in `hipMemcpy3D `?

**Comment by luraess** [comment]
From looking at the test suite,

it seems that the hip tests themselves don't test anything else than
```
myparms.srcPos = make_hipPos(0,0,0);
myparms.dstPos = make_hipPos(0,0,0);
```
as in https://github.com/ROCm-Developer-Tools/HIP/blob/22e926308893c873b7b049027727ab1700421719/tests/src/runtimeApi/memory/hipMemcpy3D.cpp#L49-L53 for the non-async version hipMemcpy3D.

And the same in the async version at https://github.com/ROCm-Developer-Tools/HIP/blob/22e926308893c873b7b049027727ab1700421719/tests/src/runtimeApi/memory/hipMemset3DRegressMultiThread.cpp#L266-L285

**Comment by jaydeeppatel1111** [comment]
Thanks @luraess, We have acknowledged issue and it will be fixed in later release.

**Comment by luraess** [comment]
> Thanks @luraess, We have acknowledged issue and it will be fixed in later release.

Thanks @jaydeeppatel1111 . Any hint on the timeline for the fix, release? Or can you provide a patch?

**Comment by mangupta** [comment]
> Thanks @jaydeeppatel1111 . Any hint on the timeline for the fix, release? Or can you provide a patch?
@luraess : The fix would make it to the next major ROCm release. On timelines for the release, it is dependent on a lot of factors and hence hard to confirm. The fix however, should make it to the develop branch of the HIP runtime repo in the next couple of weeks. In case you are in urgent need of the fix, you can then build and install the HIP runtime from the develop branch. Hope that helps answer your query.

**Comment by luraess** [comment]
>  In case you are in urgent need of the fix, you can then build and install the HIP runtime from the develop branch.

Thanks! Can you point me to where to track the landing of the fix, or ping here when done? Thanks!

**Comment by luraess** [comment]
Did this got released or is it still in the devel branch?

**Comment by jaydeeppatel1111** [comment]
Hello @luraess , Fix is available in public release 6.0 (https://github.com/ROCm/clr/tree/rocm-6.0.x).
